`std.Target`: Use `lime1` as wasm baseline model and `mvp` as generic model

alexrp opened the pull request:
See #21818 and commit messages.

(#21818 should remain open after this to track aligning `lime1` exactly with the spec when we upgrade to LLVM 20.)

alexrp commented:
This is an interesting failure: https://github.com/ziglang/zig/actions/runs/12077911317/job/33681506457?pr=22098

Only `test-behavior` exhibits this, not any of the other module tests.

alexrp commented:
Seems to be a bug with `bulk_memory` in the self-hosted wasm backend.

alexrp commented:
Ok, the `memory.copy`/`memory.fill` bug has been fixed. This should be ready for review.

alexrp commented on build.zig:
```diff
@@ -592,7 +592,10 @@ fn addWasiUpdateStep(b: *std.Build, version: [:0]const u8) !void {
         .os_tag = .wasi,
     };
-    target_query.cpu_features_add.addFeature(@intFromEnum(std.Target.wasm.Feature.bulk_memory));
+    // Not supported by the `wasm-opt` version in CI.
+    target_query.cpu_features_add.removeFeature(@intFromEnum(std.Target.wasm.Feature.extended_const));
+    // Not supported by `wasm2c`.
+    target_query.cpu_features_add.removeFeature(@intFromEnum(std.Target.wasm.Feature.nontrapping_fptoint));
```
@andrewrk would it be possible to update the `wasm-opt` version on the CI machines? I guess it would also be a good opportunity to update Wasmtime (latest is 27.0.0 vs our current 10.0.2).

andrewrk replied:
wasm-opt is coming from apt, and it would be pretty convenient to leave it that way. It's version 108 coming from Debian bookworm. If the build script is going to have a system dependency for doing update-zig1, then I think it's a reasonable limitation that the most recent stable release of debian should be good enough. I glanced at the spec for extended_const and I highly doubt it will do anything to improve zig1.wasm binary size or perf.

wasmtime on the other hand I went ahead and dropped in version v29.0.0 on all the linux machines.

andrewrk commented on src/arch/wasm/CodeGen.zig:
```diff
@@ -1591,10 +1591,28 @@ fn memcpy(cg: *CodeGen, dst: WValue, src: WValue, len: WValue) !void {
+        // Even if `len` is zero, the spec requires an implementation to trap if `src + len` or
+        // `dst + len` are out of memory bounds. This can easily happen in Zig in a case such as:
+        //
+        // const dst: *u8 = undefined;
+        // const src: *u8 = undefined;
+        // var len: usize = runtime_zero();
+        // @memcpy(dst[len..], src[len..]);
```
```suggestion
        // const dst: [*]u8 = undefined;
        // const src: [*]u8 = undefined;
        // var len: usize = runtime_zero();
        // @memcpy(dst[0..len], src[0..len]);
```

andrewrk commented on src/arch/wasm/CodeGen.zig:
```diff
@@ -4783,10 +4801,27 @@ fn memset(cg: *CodeGen, elem_ty: Type, ptr: WValue, len: WValue, value: WValue)
+        // Even if `len` is zero, the spec requires an implementation to trap if `ptr + len` is
+        // out of memory bounds. This can easily happen in Zig in a case such as:
+        //
+        // const ptr: *u8 = undefined;
+        // var len: usize = runtime_zero();
+        // @memset(ptr[len..], 42);
```
```suggestion
        // const ptr: [*]u8 = undefined;
        // var len: usize = runtime_zero();
        // @memset(ptr[0..len], 42);
```

andrewrk reviewed (approved):
Thanks for looking into this.

Related commits:

1c9bb6a79de3ca52d819177fea32ce7993634e31
836f9fceab03c7de56eba7a9c2e810206e7e8469

andrewrk commented:
What do you think about [my idea](https://github.com/ziglang/zig/issues/15920#issuecomment-1572681366) to introduce a CPU feature called `memcpy_len0_allowed` which indicates the wasm runtime always allows length zero bulk memory operations and treats them as no-ops? This would *at least* be used by zig1.wasm since we control the wasm2c implementation, and beyond that would be a small form of protest against the spec.

alexrp commented:
> What do you think about [my idea](https://github.com/ziglang/zig/issues/15920#issuecomment-1572681366) to introduce a CPU feature called `bulk_memory_len0` which indicates the wasm runtime always allows length zero bulk memory operations and treats them as no-ops? This would *at least* be used by zig1.wasm since we control the wasm2c implementation, and beyond that would be a small form of protest against the spec.

Seems reasonable. I'll add that.

alexrp commented:
Ok, I've implemented a `nontrapping_bulk_memory_len0` feature (named thusly to follow `nontrapping_fptoint`'s precedent), and also implemented `nontrapping_fptoint` support in wasm2c. I believe wasm2c is now fully `lime1`-ready; only `extended_const` is disabled for now due to wasm-opt support.

I did a `zig1.wasm` update followed by a stage3 bootstrap locally as a smoke test.

andrewrk reviewed (approved):
Looks great, thanks!

andrewrk commented on src/link/Wasm.zig:
```diff
@@ -2836,14 +2836,44 @@ pub const Feature = packed struct(u8) {
 
         pub fn fromCpuFeature(feature: std.Target.wasm.Feature) Tag {
-            return @enumFromInt(@intFromEnum(feature));
+            return switch (feature) {
```
Good move, this looks a lot better. Sorry about the trouble, I should have done this from the beginning.
